lib/cache/memory-cache-store.js

```diff
@@ -0,0 +1,314 @@
+    }
+    if (typeof value !== 'object') {
+      throw new TypeError(`expected value to be object, got ${typeof value}`)
+    }
+
+    if (this.isFull) {
+      return undefined
+    }
+
+    // TODO see if value already exists or not
+
+    const values = this.#getValues(req, true)
+    this.entryCount++
+
+    /**
+     * TODO what's a better name for this
+     * @type {MemoryStoreValue}
+     */
+    const storedValue = {
+      complete: false,
+      value,
+      emitter: new EventEmitter()
+    }
+
+    if (
+      values.length === 0 ||
+      value.deleteAt < values[values.length - 1].deleteAt
+    ) {
+      values.push(storedValue)
+    } else if (value.deleteAt >= values[0].deleteAt) {
+      values.unshift(storedValue)
+    } else {
+      let startIndex = 0
+      let endIndex = values.length
+      while (true) {
+        if (startIndex === endIndex) {
+          values.splice(startIndex, 0, storedValue)
+          break
+        }
+
+        const middleIndex = (startIndex + endIndex) / 2
+        const middleValue = values[middleIndex]
+        if (value.deleteAt === middleIndex) {
+          values.splice(middleIndex, 0, storedValue)
+          break
+        } else if (value.deleteAt > middleValue.deleteAt) {
+          endIndex = middleIndex
+          continue
+        } else {
+          startIndex = middleIndex
+          continue
+        }
+      }
+    }
+
+    return new MemoryStoreWritableStream(storedValue)
+  }
+
+  /**
+   * @param {string} origin
+   */
+  deleteByOrigin (origin) {
+    this.#data.delete(origin)
+  }
+
+  /**
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @returns {MemoryStoreValue[] | undefined}
+   */
+  #getValues (req, makeIfDoesntExist) {
+    // https://www.rfc-editor.org/rfc/rfc9111.html#section-2-3
+    let cachedPaths = this.#data.get(req.origin)
+    if (!cachedPaths) {
+      if (!makeIfDoesntExist) {
+        return undefined
+      }
+
+      cachedPaths = new Map()
+      this.#data.set(req.origin, cachedPaths)
+    }
+
+    let values = cachedPaths.get(`${req.path}:${req.method}`)
+    if (!values && makeIfDoesntExist) {
+      values = []
+      cachedPaths.set(`${req.path}:${req.method}`, values)
+    }
+
+    return values
+  }
+}
+
+class MemoryStoreReadableStream extends Readable {
+  /**
+   * @type {MemoryStoreValue}
+   */
+  #value
+  /**
+   * @type {Buffer[]}
+   */
+  #chunksToSend
+
+  /**
+   * @param {MemoryStoreValue} value
+   */
+  constructor (value) {
+    super()
+
+    this.#value = value
+    this.#chunksToSend = [...this.#value.value.body]
+
+    if (!value.complete) {
+      value.emitter.on('write', this.#chunksToSend.push)
+      value.emitter.on('final', () => {
+        this.#chunksToSend.push(null)
+      })
+    }
+  }
+
+  get value () {
+    return this.#value
+  }
+
+  /**
+   * @param {number} size
+   */
+  _read (size) {
+    // TODO what to do if we get a read but we're still waiting on more chunks?
```

Comment:
not sure what we could do here

Reply:
The stream protocol handles this. Just return.

How is the backpressure handled? In other terms, is it possible that a slow consumer of this can slow down the producers of the chunks?

I don't think we can, so you'll accumulate all the chunks in memory anyway. Having streams in between is not bad anyway, as it will allow streamed cached response.